ref-arch/real-time-analytics.rst

```diff
@@ -0,0 +1,509 @@
+      RETURN;
+    END IF;
+    PERFORM RESET lock_timeout;
+
+    EXECUTE format('
+      WITH (
+        INSERT INTO %I (
+            zone_id, ingest_time, request_count,
+            error_count, success_count, average_response_time_msec)
+          SELECT
+            zone_id,
+            date_trunc('minute', ingest_time) as minute,
+            COUNT(1) as request_count,
+            COUNT(CASE WHEN (status_code between 200 and 299) THEN 1 ELSE 0 END) as success_count,
+            request_count - success_count AS error_count,
+            SUM(response_time_msec) / COUNT(1) AS average_response_time_msec
+            -- later sections will show some more clauses that can go here
+          FROM %I
+          WHERE ingest_time > v_latest_minute_already_aggregated
+          GROUP BY zone_id, minute
+        ) as inserted_rows
+      SELECT max(minute) INTO v_new_latest_already_aggregated FROM inserted_rows;
+    ', dest_shard, source_shard);
+
+    -- mark how much work we did, so the next invocation picks up where we left off
+    PERFORM UPDATE rollup_thresholds
+      SET ingest_time = v_new_latest_already_aggregated
+      WHERE source_shard = source_shard AND dest_shard = dest_shard;
+  END;
+  $$ LANGUAGE 'plpgsql';
+
+As discussed above, there's a 1-to-1 correspondence between http_request shards and
+``http_request_1min`` shards. This function accepts the name of the ``http_request`` shard
+to read from and the name of the ``http_request_1min`` shard to write to. It can't figure
+it out itself because that kind of metadata is kept on the master, not the workers.
+
+It also uses a local table, to keep track of how much of the raw data has already been
+aggregated:
+
+.. code-block:: sql
+
+  -- every worker should have their own local version of this table
+  CREATE TABLE rollup_thresholds (
+        ingest_time timestamptz,
+        source_shard regclass,
+        dest_shard regclass,
+        UNIQUE (source_shard, dest_shard)
+  );
+
+Since this function is given some metadata from the master, where does the master get that
+metadata from? Every minute it calls its own function which fires off all the
+aggregations:
+
+.. code-block:: plpgsql
+
+  -- this should be run on the master
+  CREATE FUNCTION run_rollups(source_table text, dest_table text) RETURNS void
+  AS $$
+  DECLARE
+  BEGIN
+    FOR source_shard, dest_shard, nodename, nodeport IN
+      SELECT
+        a.logicalrelid::regclass||'_'||a.shardid,
+        b.logicalrelid::regclass||'_'||b.shardid,
+        nodename, nodeport
+      FROM pg_dist_shard a
+      JOIN pg_dist_shard b USING (shardminvalue)
+      JOIN pg_dist_shard_placement p ON (a.shardid = p.shardid)
+      WHERE a.logicalrelid = 'first'::regclass AND b.logicalrelid = 'second'::regclass;
+    LOOP
+      SELECT * FROM dblink(
+        format('host=%s port=%d', nodename, nodeport),
+        format('SELECT rollup_1min(%, %s);', source_shard, dest_shard));
```

Comment:
dblink requires a list of output columns, which works poorly with calling void functions.

You could make rollup_1min return timestamptz and put "AS (out timestamptz)" behind the dblink call.